File: branches/feature/iPhone_group_toggle_polling/Classes/SwitchView.m
/* OpenRemote, the Home of the Digital Home.
 * Copyright 2008-2009, OpenRemote Inc.
 * 
 * See the contributors.txt file in the distribution for a
 * full listing of individual contributors.
 * 
 * This is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3.0 of
 * the License, or (at your option) any later version.
 * 
 * This software is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this software; if not, write to the Free
 * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
 * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
 */

#import "SwitchView.h"
#import "DirectoryDefinition.h"
#import "ViewHelper.h"
#import "PollingStatusParserDelegate.h"
#import "NotificationConstant.h"


@interface SwitchView (Private)


- (void)setOn:(BOOL)on;
- (void)stateChanged:(id)sender;

@end



@implementation SwitchView

@synthesize button, onUIImage, offUIImage;

- (void)stateChanged:(id)sender {
	if (isOn) {
		[self sendCommandRequest:@"OFF"];
	} else {		
		[self sendCommandRequest:@"ON"];
	} 
}

- (void)setPollingStatus:(NSNotification *)notification {
	PollingStatusParserDelegate *delegate = (PollingStatusParserDelegate *)[notification object];
	NSString *newStatus = [delegate.statusMap objectForKey:[NSString stringWithFormat:@"%d",control.controlId]];
	if ([[newStatus uppercaseString] isEqualToString:@"ON"]) {
		[self setOn:YES];
	} else if ([[newStatus uppercaseString] isEqualToString:@"OFF"]) {
		[self setOn:NO];
	} 
}


- (void)setOn:(BOOL)on {
	if (on) {
		isOn = YES;
		if (canUseImage) {
			[button setImage:onUIImage forState:UIControlStateNormal];		
		} else {
			[button setTitle:@"ON" forState:UIControlStateNormal];			
		}

	} else {
		isOn = NO;
		if (canUseImage) {
			[button setImage:offUIImage forState:UIControlStateNormal];			
		} else {
			[button setTitle:@"OFF" forState:UIControlStateNormal];
		}		
	}

}



//Create button according to control and add tap event
- (void)createButton {
	
	//avoid overlap the same button by mistake
	if (button) {
		[button removeFromSuperview];
		[button release];
	}
	
	button = [[UIButton buttonWithType:UIButtonTypeCustom] retain];
	[button addTarget:self action:@selector(stateChanged:) forControlEvents:UIControlEventTouchUpInside];
	
	[self addSubview:button];
	Switch *theSwitch = (Switch *)control;
	NSString *onImage = theSwitch.onImage.src;
	NSString *offImage = theSwitch.offImage.src;
	
	//assign YES if have both on image and off image
	canUseImage = onImage && offImage;
}

//NOTE:You should init all nested views with *initWithFrame* and you should pass in valid frame rects.
//Otherwise, UI widget inside will not work in nested UIViews
- (void)layoutSubviews {
	[self createButton];
	Switch *theSwitch = (Switch *)control;
	NSString *onImage = theSwitch.onImage.src;
	NSString *offImage = theSwitch.offImage.src;
	[button setFrame:[self bounds]];
	if (canUseImage) {
		 onUIImage = [[UIImage alloc] initWithContentsOfFile:[[DirectoryDefinition imageCacheFolder] stringByAppendingPathComponent:onImage]];
		 offUIImage = [[UIImage alloc] initWithContentsOfFile:[[DirectoryDefinition imageCacheFolder] stringByAppendingPathComponent:offImage]];
	 } else {
		UIImage *buttonImage = [[UIImage imageNamed:@"button.png"] stretchableImageWithLeftCapWidth:20 topCapHeight:20];
		[button setBackgroundImage:buttonImage forState:UIControlStateNormal];

		buttonImage = [[UIImage imageNamed:@"buttonHighlighted.png"] stretchableImageWithLeftCapWidth:20 topCapHeight:20];
		[button setBackgroundImage:buttonImage forState:UIControlStateHighlighted];

		button.titleLabel.font = [UIFont boldSystemFontOfSize:18];
		[button setTitleShadowColor:[UIColor grayColor] forState:UIControlStateNormal];
		button.titleLabel.shadowOffset = CGSizeMake(0, -2);
	}
	[self setOn:NO];
	[[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(setPollingStatus:) name:[NSString stringWithFormat:NotificationPollingStatusIdFormat,control.controlId] object:nil];
}



- (void)dealloc {
	[button release];
	[onUIImage release];
	[offUIImage release];
	[super dealloc];
}


@end

File: workspace/tchougbegrace/Android/src/org/openremote/android/console/view/SwitchView.java
/* OpenRemote, the Home of the Digital Home.
* Copyright 2008-2010, OpenRemote Inc.
*
* See the contributors.txt file in the distribution for a
* full listing of individual contributors.
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU Affero General Public License as
* published by the Free Software Foundation, either version 3 of the
* License, or (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
* GNU Affero General Public License for more details.
*
* You should have received a copy of the GNU Affero General Public License
* along with this program. If not, see <http://www.gnu.org/licenses/>.
*/
package org.openremote.android.console.view;

import org.openremote.android.console.Constants;
import org.openremote.android.console.bindings.Switch;
import org.openremote.android.console.model.ListenerConstant;
import org.openremote.android.console.model.OREvent;
import org.openremote.android.console.model.OREventListener;
import org.openremote.android.console.model.ORListenerManager;
import org.openremote.android.console.model.PollingStatusParser;
import org.openremote.android.console.util.ImageUtil;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.os.Handler;
import android.os.Message;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;

/**
 * This class is responsible for rendering the switch in screen with the switch data.
 * It has control command and sensory.
 * 
s */
public class SwitchView extends SensoryControlView {

   private Button button;
   private Drawable onImage;
   private Drawable offImage;
   private boolean isOn;
   private boolean canUseImage;
   public SwitchView(Context context, Switch switchComponent) {
      super(context);
      setComponent(switchComponent);
      if (switchComponent != null) {
         button = new Button(context, null, android.R.attr.buttonStyleSmall);
         button.setTextSize(Constants.DEFAULT_FONT_SIZE);
         initSwitch(switchComponent);
         if (switchComponent.getSensor() != null) {
            addPollingSensoryListener();
         }
      }
      setLayoutParams(new FrameLayout.LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT));
   }

   /**
    * Sets switch's on/off state(image or text).
    * Sets touch listener on the switch view for change switch state and send command.
    * 
    * @param switchComponent the switch component
    */
   private void initSwitch(Switch switchComponent) {
	  
      int width = switchComponent.getFrameWidth();
      int height = switchComponent.getFrameHeight();
      button.setLayoutParams(new FrameLayout.LayoutParams(width, height));
      if (switchComponent.getOnImage() != null) {
         onImage = ImageUtil.createFromPathQuietly(Constants.FILE_FOLDER_PATH + switchComponent.getOnImage().getSrc());
         button.setLayoutParams(new FrameLayout.LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.WRAP_CONTENT));
      }
      if (switchComponent.getOffImage() != null) {
         offImage = ImageUtil.createFromPathQuietly(Constants.FILE_FOLDER_PATH + switchComponent.getOffImage().getSrc());
      }
      if (onImage != null && offImage != null) {
         canUseImage = true;
         button.setText(null);
         if (isOn) {
            button.setBackgroundDrawable(onImage);
         } else {
            button.setBackgroundDrawable(offImage);
         }
      } else {
         if (isOn) {
            button.setText("ON");
         } else {
            button.setText("OFF");
         }
      }
      button.setOnTouchListener(new OnTouchListener() {
    	  
         public boolean onTouch(View v, MotionEvent event) {
            if (event.getAction() == MotionEvent.ACTION_DOWN) {
               if (canUseImage) {
                  if (isOn) {
                     onImage.setAlpha(200);
                     button.setBackgroundDrawable(onImage);
                  } else {
                     offImage.setAlpha(200);
                     button.setBackgroundDrawable(offImage);
                  }
               }
            } else if (event.getAction() == MotionEvent.ACTION_UP) {
               if (canUseImage) {
                  if (isOn) {
                     onImage.setAlpha(255);
                     button.setBackgroundDrawable(onImage);
                  } else {
                     offImage.setAlpha(255);
                     button.setBackgroundDrawable(offImage);
                  }
               }
               if (isOn) {
                  sendCommandRequest(Switch.OFF);
               } else {
                  sendCommandRequest(Switch.ON);
               }
            }
            return false;
         }
        
     });
      addView(button);
   }
   
   @Override
   public void addPollingSensoryListener() {
      final Integer sensorId = ((Switch)getComponent()).getSensor().getSensorId();
      if (sensorId > 0) {
         ORListenerManager.getInstance().addOREventListener(ListenerConstant.ListenerPollingStatusIdFormat + sensorId, new OREventListener() {
            public void handleEvent(OREvent event) {
               String value = PollingStatusParser.statusMap.get(sensorId.toString()).toLowerCase();
               if (isOn && Switch.OFF.equals(value)) {
                  isOn = false;
               } else if (!isOn && Switch.ON.equals(value)) {
                  isOn = true;
               }
               handler.sendEmptyMessage(0);
            }
         });
      }
   }
   
   /** The handler is for updating switch state by polling result. */
   private Handler handler = new Handler() {
      @Override
      public void handleMessage(Message msg) {
         if (canUseImage) {
            if (isOn) {
               button.setBackgroundDrawable(onImage);
            } else {
               button.setBackgroundDrawable(offImage);
            }
         } else {
            if (isOn) {
               button.setText("ON");
            } else {
               button.setText("OFF");
            }
         }
         super.handleMessage(msg);
      }
  };
  
}
